pyquil/api/_config.py

```diff
@@ -226,6 +230,16 @@ def dispatch_url(self) -> str:
     def engage_cmd(self) -> str:
         return self._env_or_config_or_default(**self.ENGAGE_CMD)
 
+    @property
+    def engagement(self) -> Optional['Engagement']:
+        """
+        An Engagement should only be made available to consumers if it was used to retrieve a
+        configuration value.
+        """
+        if not self._engagement_requested:
+            return
+        return self.get_engagement()
```

Comment:
do we actually need this property? why can't this branching be in `get_engagement`?

Reply:
because the job of `ForestSession#get_engagement` is to always get an engagement when it's asked to. Config has to "remember" if engagement was ever needed.